server/snapshot.go

```diff
@@ -234,6 +246,27 @@ func (r *raftFSM) Restore(snapshot io.ReadCloser) (retErr error) {
 	r.Lock()
 	defer r.Unlock()
 
+	shouldSnapshot := false
+	defer func() {
+		if retErr == nil && shouldSnapshot {
+			s.wg.Add(1)
+			go func() {
+				defer s.wg.Done()
+				// s.raft.Snapshot() is actually s.raft.Raft.Snapshot()
+				// (our raft structure embeds Raft). But it is set after
+				// the call to NewRaft() - which invokes this function on
+				// startup. However, this runs under the server's lock, so
+				// simply grab the lock to ensure that Snapshot() will be
+				// able to execute.
+				// Also don't use startGoRoutine() here since it needs the
+				// server lock, which we already have.
+				s.mu.Lock()
+				s.raft.Snapshot().Error()
```

Comment:
Can't do in place correct?

Reply:
meaning without go routine? if so no, since as I tried to explain, as part of the startup process we call NewRaft() (hashicorp raft constructor), which returns a Raft object that we assign to s.raft.Raft at the end of createRaftNode(). NewRaft() is the one invoking Restore() on startup (when there are local snapshots), so I can't call s.raft.Snapshot() until after the creation of the raft node has completed.